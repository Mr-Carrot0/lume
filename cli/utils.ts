import lume from "../mod.ts";
import { log } from "../core/utils/log.ts";
import Site from "../core/site.ts";

/** Create a site instance */
export async function createSite(_config?: URL): Promise<Site> {
  if (!_config) {
    return lume();
  }
  log.info(`Loading config file <gray>${_config}</gray>`);
  const mod = await import(_config.toString());
  const site = mod.default as Site | undefined;
  if (!(site instanceof Site)) {
    log.fatal(
      `[Lume] Missing Site instance! Ensure your config file does export the Site instance as default.`,
    );
    throw new Error("Site instance is not found");
  }
  return site;
}

/** Create a site intance and build it */
export async function buildSite(site: Site): Promise<void> {
  performance.mark("start");
  await site.build();
  performance.mark("end");

  log.info(`üçæ Site built into <gray>${site.options.dest}</gray>`);
  const duration = performance.measure("duration", "start", "end").duration /
    1000;
  const total = site.pages.length + site.files.length;
  log.info(
    `  <gray>${total} files generated in ${duration.toFixed(2)} seconds</gray>`,
  );
}
