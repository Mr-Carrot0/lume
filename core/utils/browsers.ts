/**
 * Code generated from:
 *
 * ```ts
 * import { getCompatibleVersions } from "npm:baseline-browser-mapping@2.4.4";
 *
 * const results = getCompatibleVersions({
 *   widelyAvailableOnDate: "2025-05-07", // Lume 3.0.0 release date
 * });
 * ```
 *
 * Not run directly to avoid importing this package that requires read access.
 */
export const browsers = {
  chrome: [107, 0],
  chrome_android: [107, 0],
  edge: [107, 0],
  firefox: [104, 0],
  firefox_android: [104, 0],
  safari: [16, 0],
  safari_ios: [16, 0],
} as Record<string, [number, number]>;

export function version(browser: [number, number]): number {
  return (browser[0] << 16) | (browser[1] << 8);
}
export function versionString(browser: [number, number]): string {
  return `${browser[0]}.${browser[1]}`;
}
