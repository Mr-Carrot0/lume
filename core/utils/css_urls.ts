type UrlWalker = (url: string, type: string) => string | Promise<string>;
type Position = [string, string, number, number];
type Index = [string, number];

export async function walkUrls(
  code: string,
  walker: UrlWalker,
): Promise<string> {
  let result = "";
  let index = 0;
  let position = next(code, index);

  while (position) {
    const [type, url, start, end] = position;
    const newUrl = await walker(url, type);
    const quotedUrl = url.startsWith("data:")
      ? `'${newUrl.replaceAll("'", "\\'")}'`
      : `"${newUrl.replaceAll('"', '\\"')}"`;
    result += `${code.slice(index, start)}${quotedUrl}`;
    index = end;
    position = next(code, index);
  }

  return result + code.slice(index);
}

function next(code: string, index = 0): Position | undefined {
  const nextUrl = code.indexOf("url(", index);
  const nextComment = code.indexOf("/*", index);
  const nextImport = code.indexOf("@import ", index);

  let found: Index | undefined;

  if (nextUrl !== -1) {
    found = ["url", nextUrl];
  }
  if (nextComment !== -1 && (!found || nextComment < found[1])) {
    found = ["comment", nextComment];
  }
  if (nextImport !== -1 && (!found || nextImport < found[1])) {
    found = ["import", nextImport];
  }

  if (!found) {
    return;
  }

  // If it's a comment, find the end of it
  if (found[0] === "comment") {
    const end = code.indexOf("*/", found[1]);
    if (end === -1) {
      return;
    }
    return next(code, end + 2);
  }

  if (found[0] === "import") {
    const double = code.indexOf('"', found[1]);
    const single = code.indexOf("'", found[1]);
    const urlFn = code.indexOf("url(", found[1]);

    let foundImport: Index | undefined;

    if (double !== -1) {
      foundImport = ["double", double];
    }
    if (single !== -1 && (!foundImport || single < foundImport[1])) {
      foundImport = ["single", single];
    }

    // It's has a url(), so delegate to the url() logic
    if (urlFn !== -1 && (!foundImport || urlFn < foundImport[1])) {
      return next(code, urlFn);
    }

    if (!foundImport) {
      return;
    }

    const start = foundImport[1];
    const end =
      code.indexOf(foundImport[0] === "double" ? '"' : "'", start + 1) + 1;
    return ["url", code.slice(start + 1, end - 1), start, end];
  }

  if (found[0] === "url") {
    const start = code.indexOf("(", found[1]) + 1;
    let url = code.slice(start).trim();
    const quote = url.startsWith('"') || url.startsWith("'")
      ? url[0]
      : undefined;
    url = quote ? url.slice(1) : url;
    let end = findUrlEnd(url, quote);
    url = url.slice(0, end);
    end += start + (quote ? 2 : 0);
    url = url.replace(/\\'/g, "'").replace(/\\"/g, '"');
    return ["url", url, start, end];
  }
}

function findUrlEnd(code: string, quote?: string): number {
  if (!quote) {
    return code.indexOf(")");
  }

  let end = code.indexOf(quote);

  while (end !== -1 && code[end - 1] === "\\") {
    end = code.indexOf(quote, end + 1);
  }

  return end === -1 ? code.indexOf(")") : end;
}
