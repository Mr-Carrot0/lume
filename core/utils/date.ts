/**
 * Returns the date of the git commit that created or modified the file.
 * Thanks to https://github.com/11ty/eleventy/blob/8dd2a1012de92c5ee1eab7c37e6bf1b36183927e/src/Util/DateGitLastUpdated.js
 */
export function getGitDate(
  type: "created" | "modified",
  file: string,
): Date | undefined {
  const args = type === "created"
    ? ["log", "--diff-filter=A", "--follow", "-1", "--format=%at", "--", file]
    : ["log", "-1", "--format=%at", "--", file];

  const { stdout, success } = new Deno.Command("git", { args }).outputSync();

  if (!success) {
    return;
  }
  const str = new TextDecoder().decode(stdout);

  if (str) {
    return parseDate(parseInt(str) * 1000);
  }
}

/** Parse a string or number (of miliseconds) to UTC Date */
export function parseDate(date: string | number): Date {
  return new Date(getZonedDateTime(date).epochMilliseconds);
}

/** Parse a string or number (of miliseconds) to a zoned datetime */
export function getZonedDateTime(
  date: string | number,
  timezone = "UTC",
): Temporal.ZonedDateTime {
  if (typeof date === "number") {
    return Temporal.Instant.fromEpochMilliseconds(date).toZonedDateTimeISO(
      timezone,
    );
  }

  try {
    return Temporal.Instant.from(date).toZonedDateTimeISO(timezone);
  } catch {
    return Temporal.PlainDateTime.from(date).toZonedDateTime(timezone);
  }
}
