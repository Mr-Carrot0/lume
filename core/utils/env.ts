const envVars = new Map<string, string>();

export function setEnv(name: string, value: string) {
  if (allowedEnvVars()) {
    Deno.env.set(name, value);
  }
  envVars.set(name, value);
}

export function env<T>(name: string): T | undefined {
  const value = allowedEnvVars()
    ? envVars.get(name) ?? Deno.env.get(name)
    : envVars.get(name);

  if (typeof value === "undefined") {
    return undefined;
  }

  switch (value.toLowerCase()) {
    case "true":
    case "on":
    case "1":
      return true as T;

    case "false":
    case "off":
    case "0":
      return false as T;

    default:
      return value as T;
  }
}

let allowed: boolean | undefined;
function allowedEnvVars(): boolean {
  if (allowed === undefined) {
    allowed = Deno.permissions.querySync?.({ name: "env" }).state === "granted";
  }
  return allowed;
}
