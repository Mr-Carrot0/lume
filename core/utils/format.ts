const byteUnit = Intl.NumberFormat("en", {
  notation: "compact",
  style: "unit",
  unit: "byte",
  unitDisplay: "narrow",
});

export function bytes(bytes: number) {
  return byteUnit.format(bytes);
}

const percentageUnit = Intl.NumberFormat("en", {
  style: "percent",
  minimumFractionDigits: 1,
  maximumFractionDigits: 1,
});

export function percentage(initialValue: number, finalValue: number) {
  if (initialValue === finalValue) {
    return "0%";
  }

  const ratio = -(1 - finalValue / initialValue);

  if (ratio < 0) {
    return percentageUnit.format(ratio);
  }

  return "+" + percentageUnit.format(ratio);
}

const secondUnit = Intl.NumberFormat("en", {
  notation: "compact",
  style: "unit",
  unit: "second",
  unitDisplay: "short",
  minimumFractionDigits: 1,
  maximumFractionDigits: 1,
});

const milisecondUnit = Intl.NumberFormat("en", {
  notation: "compact",
  style: "unit",
  unit: "millisecond",
  unitDisplay: "short",
  maximumFractionDigits: 0,
});

export function duration(miliseconds: number) {
  if (miliseconds < 1000) {
    return milisecondUnit.format(miliseconds);
  }

  return secondUnit.format(miliseconds / 1000);
}
