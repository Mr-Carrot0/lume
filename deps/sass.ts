export * from "npm:sass@1.88.0";
