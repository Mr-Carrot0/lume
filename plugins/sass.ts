import {
  normalizePath,
  replaceExtension,
  resolveInclude,
} from "../core/utils/path.ts";
import { merge } from "../core/utils/object.ts";
import { concurrent } from "../core/utils/concurrent.ts";
import { compileStringAsync } from "../deps/sass.ts";
import { fromFileUrl, posix, toFileUrl } from "../deps/path.ts";
import { Page } from "../core/file.ts";
import { prepareAsset, saveAsset } from "./source_maps.ts";
import { log } from "../core/utils/log.ts";

import type Site from "../core/site.ts";
import type { StringOptions } from "../deps/sass.ts";

type SassOptions = Omit<StringOptions<"async">, "url" | "syntax">;

export interface Options {
  /** Output format */
  format?: "compressed" | "expanded";

  /**
   * SASS options
   * @see https://sass-lang.com/documentation/js-api/interfaces/options/
   */
  options?: SassOptions;

  /**
   * Custom includes paths
   * @default `site.options.includes`
   */
  includes?: string;
}

export const defaults: Options = {
  format: "compressed",
  options: {},
};

/**
 * A plugin to process SASS and SCSS files
 * @see https://lume.land/plugins/sass/
 */
export function sass(userOptions?: Options) {
  return (site: Site) => {
    const options = merge(
      { ...defaults, includes: site.options.includes },
      userOptions,
    );

    // Ignore includes folder
    if (options.includes) {
      site.ignore(options.includes);
    }

    site.process([".scss", ".sass"], sassProcessor);

    function sassProcessor(files: Page[]) {
      if (files.length === 0) {
        log.warn(
          "[sass plugin] No SCSS or SASS files found. Make sure to add them with <gray>site.add()</gray>",
        );
        return;
      }

      return concurrent(files, sass);
    }

    const { entries } = site.fs;
    const basePath = site.src();

    async function sass(page: Page) {
      const { content, filename, enableSourceMap } = prepareAsset(site, page);
      const baseFilename = posix.dirname(filename);

      const sassOptions: StringOptions<"async"> = {
        ...options.options,
        sourceMap: enableSourceMap,
        style: options.format,
        syntax: page.src.ext === ".sass" ? "indented" : "scss",
        url: toFileUrl(filename),
        importer: {
          canonicalize(url: string) {
            const pathname = normalizePath(fromFileUrl(url));
            const mainPath = pathname.startsWith(basePath)
              ? normalizePath(pathname.slice(basePath.length))
              : pathname;

            for (const path of getPathsToLook(mainPath)) {
              const entry = entries.get(path);

              if (entry) {
                return toFileUrl(site.src(entry.path));
              }
            }

            // Search in includes
            const includePath = pathname.startsWith(baseFilename)
              ? pathname.slice(baseFilename.length)
              : mainPath;

            for (const path of getPathsToLook(includePath)) {
              const resolved = resolveInclude(path, options.includes);
              const entry = entries.get(resolved);

              if (entry) {
                return toFileUrl(site.src(entry.path));
              }
            }

            return null;
          },
          async load(url: URL) {
            const pathname = fromFileUrl(url);
            const contents = await site.getContent(pathname, false);

            if (typeof contents === "string") {
              return {
                contents,
                syntax: pathname.endsWith(".sass") ? "indented" : "scss",
                sourceMapUrl: url,
              };
            }

            throw new Error(`File not found: ${url} (${pathname})`);
          },
        },
      };

      const output = await compileStringAsync(content, sassOptions);

      // @ts-ignore: sourceMap is not in the type definition
      saveAsset(site, page, output.css, output.sourceMap);
      page.data.url = replaceExtension(page.data.url, ".css");
    }
  };
}

function getPathsToLook(path: string): string[] {
  const basename = posix.basename(path);

  if (posix.extname(path)) {
    if (basename.startsWith("_")) {
      return [path];
    }
    return [path, posix.join(posix.dirname(path), `_${basename}`)];
  } else {
    if (basename.startsWith("_")) {
      return [
        `${path}.scss`,
        `${path}.sass`,
      ];
    }

    return [
      `${path}.scss`,
      `${path}.sass`,
      posix.join(posix.dirname(path), `_${basename}.scss`),
      posix.join(posix.dirname(path), `_${basename}.sass`),
      posix.join(path, `index.scss`),
      posix.join(path, `index.sass`),
      posix.join(path, `_index.scss`),
      posix.join(path, `_index.sass`),
    ];
  }
}

export default sass;
