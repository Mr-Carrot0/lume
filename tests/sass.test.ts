import { assertSiteSnapshot, build, getSite } from "./utils.ts";
import sass from "../plugins/sass.ts";

Deno.test("SASS plugin", async (t) => {
  const site = getSite({
    src: "sass",
  });
  site.add([".scss", ".sass"]);
  site.use(sass());

  await build(site);
  await assertSiteSnapshot(t, site);
});
