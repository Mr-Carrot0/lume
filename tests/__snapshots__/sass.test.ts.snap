export const snapshot = {};

snapshot[`sass plugin 1`] = `
{
  formats: [
    {
      engines: 0,
      ext: ".page.toml",
      isPage: true,
      loader: [AsyncFunction: toml],
    },
    {
      engines: 1,
      ext: ".page.ts",
      isPage: true,
      loader: [AsyncFunction: module],
    },
    {
      engines: 1,
      ext: ".page.js",
      isPage: true,
      loader: [AsyncFunction: module],
    },
    {
      engines: 0,
      ext: ".page.jsonc",
      isPage: true,
      loader: [AsyncFunction: json],
    },
    {
      engines: 0,
      ext: ".page.json",
      isPage: true,
      loader: [AsyncFunction: json],
    },
    {
      dataLoader: [AsyncFunction: json],
      engines: 0,
      ext: ".json",
      loader: [AsyncFunction: json],
    },
    {
      dataLoader: [AsyncFunction: json],
      engines: 0,
      ext: ".jsonc",
      loader: [AsyncFunction: json],
    },
    {
      engines: 1,
      ext: ".md",
      isPage: true,
      loader: [AsyncFunction: text],
    },
    {
      engines: 1,
      ext: ".markdown",
      isPage: true,
      loader: [AsyncFunction: text],
    },
    {
      dataLoader: [AsyncFunction: module],
      engines: 1,
      ext: ".js",
      loader: [AsyncFunction: module],
    },
    {
      dataLoader: [AsyncFunction: module],
      engines: 1,
      ext: ".ts",
      loader: [AsyncFunction: module],
    },
    {
      engines: 1,
      ext: ".vento",
      isPage: true,
      loader: [AsyncFunction: text],
    },
    {
      engines: 1,
      ext: ".vto",
      isPage: true,
      loader: [AsyncFunction: text],
    },
    {
      dataLoader: [AsyncFunction: toml],
      engines: 0,
      ext: ".toml",
      loader: [AsyncFunction: toml],
    },
    {
      dataLoader: [AsyncFunction: yaml],
      engines: 0,
      ext: ".yaml",
      isPage: true,
      loader: [AsyncFunction: yaml],
    },
    {
      dataLoader: [AsyncFunction: yaml],
      engines: 0,
      ext: ".yml",
      isPage: true,
      loader: [AsyncFunction: yaml],
    },
    {
      ext: ".scss",
    },
    {
      ext: ".sass",
    },
  ],
  src: [
    "/",
    "/_includes",
    "/_includes/sass",
    "/_includes/sass/_list.scss",
    "/_includes/sass/bootstrap.scss",
    "/_includes/sass/button",
    "/_includes/sass/button/index.scss",
    "/_includes/sass/colors.scss",
    "/_includes/sass/third.scss",
    "/index.scss",
    "/index2.sass",
    "/styles.css",
  ],
}
`;

snapshot[`sass plugin 2`] = `[]`;

snapshot[`sass plugin 3`] = `
[
  {
    content: '@import"./styles.css";html{scroll-behavior:smooth;font:100% Helvetica,sans-serif}body{font:100% Helvetica,sans-serif;color:#333}li{margin:0;padding:0;list-style:none}',
    data: {
      basename: "index",
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      url: "/index.css",
    },
    src: {
      ext: ".scss",
      path: "/index",
      remote: undefined,
    },
  },
  {
    content: "body{color:blue}",
    data: {
      basename: "index2",
      mergedKeys: [
        "tags",
      ],
      page: [
        "src",
        "data",
      ],
      paginate: "paginate",
      search: [],
      tags: "Array(0)",
      url: "/index2.css",
    },
    src: {
      ext: ".sass",
      path: "/index2",
      remote: undefined,
    },
  },
]
`;
